### ISSUE TITLE
2.8.0-RC1 Significant Performance Regression -- public static String valueOf(char data[], int offset, int count)

### ISSUE THREAD

**Issue by ainsley** [gwtproject/gwt#9383]
I have been developing an application that involves client side text parsing. I noticed a significant performance regression when compiling my heavy GWT application with 2.8.0-RC1. The slowdown was an order of magnitude slower on Chrome 10x, on Firefox 3x, and on IE 2x.

I've managed to use the Chrome debugger to narrow the location of the regression down to the new 2.8.0 implementation of "java.lang.String.valueOf(char data[], int offset, int count)".

**2.7.0 - Baseline implementation**

Here is a partial implementation of java.lang.String.valueOf(char data[], int offset, int count) in 2.7.0.

```
function java_lang_String_valueOf___3CIILjava_lang_String_2(x_0, offset, count){
  var end;
  end = offset + count;
  java_lang_String__1_1checkBounds__IIIV(x_0.length, offset, end);
  return java_lang_String__1_1valueOf___3CIILjava_lang_String_2(x_0, offset, end);
}
```
 
**2.8.0 RC1 - New Implementation - with performance regression**

Seems to have occurred in this commit (which seems to have been made to fix a chrome bug) : https://github.com/gwtproject/gwt/commit/b5240d61fbbbcc3619028eead0b89e4de2e7e400

```
function java_lang_String_valueOf___3CIILjava_lang_String_2(x_0, offset, count){
  java_lang_String_$clinit__V();
  var batchEnd, batchStart, end, s;
  end = offset + count;
  javaemul_internal_InternalPreconditions_checkCriticalStringBounds__IIIV(offset, end, x_0.length);
  s = '';
  for (batchStart = offset; batchStart < end;) {
    batchEnd = batchStart + $intern_13 < end?batchStart + $intern_13:end;
    s += java_lang_String_fromCharCode___3Ljava_lang_Object_2Ljava_lang_String_2(x_0.slice(batchStart, batchEnd));
    batchStart = batchEnd;
  }
  return s;
}
```

**Manual Performance Fix -- NOTE :: THIS REINTRODUCES THE POSSIBILITY OF STACK OVERFLOW**

My hand-coded fix is to simply return the 2.7.0 implementation whist making use of the new  'javaemul_internal_InternalPreconditions_checkCriticalStringBounds__IIIV' method, and changing the order of the parameters versus the old 'java_lang_String__1_1checkBounds__IIIV' method. I also create a supporting function that does not seem to be present in my 2.8.0 generated code:

```
// COPIED FROM 2.7.0
function java_lang_String_valueOf___3CIILjava_lang_String_2(x_0, offset, count){
  var end;
  end = offset + count;
  javaemul_internal_InternalPreconditions_checkCriticalStringBounds__IIIV(offset, end, x_0.length);
  return java_lang_String__1_1valueOf___3CIILjava_lang_String_2(x_0, offset, end);
}

// COPIED FROM 2.7.0
function java_lang_String__1_1valueOf___3CIILjava_lang_String_2(x_0, start_0, end){
  var s = '';
  for (var batchStart = start_0; batchStart < end;) {
    var batchEnd = Math.min(batchStart + 10000, end);
    s += String.fromCharCode.apply(null, x_0.slice(batchStart, batchEnd));
    batchStart = batchEnd;
  }
  return s;
}
```

**Summary**

Upon manually updating the generated (prettified) GWT 2.8.0 compiler output, I record that not only does the regression disappear, but the 2.8.0 shows a 30% speed boost over 2.7 (in Chrome). That's approx 13x faster than prior to the change. Obviously this speed boost is very specific to my own use-case (which is extremely string heavy)

I don't really know why the 2.7.0 implementation is so much faster than the 2.8.0 implementation. There are a lot of secret optimizations that occur in JavaScript engines and mastering performance is nigh on impossible. All I can say is that I tested and observed a 10x performance regression on Chrome, 3x on Firefox, and 2x on IE with the new 2.8 code, although I realise that a test harness would really help to prove the regression. I wonder if someone can put one together?

Chris

**Comment by ainsley** [comment]
Investigating this a bit more. It appears the code was changed in order to fix a bug where if a large buffer was provided, Chrome throw a strack overflow exception. An alternative implementation was written that splits the operation into smaller batches so that there is not stack overflow.

Early in the history of this change it appears that there was a call for a hybrid approach see here https://gwt-review.googlesource.com/#/c/6310/1/user/super/com/google/gwt/emul/java/lang/String.java ). That is, if the active part of the buffer is less than a certain size (10,000 was suggested), then use the old approach, otherwise use the batch based approach.

Now, I don't know why this approach was dismissed, but it appears that the new batch based approach is significantly slower for smaller buffer sizes. But of course, the bug still needs to be fixed. I would suggest that the hybrid based approach be evaluated and revisited.

This may seem like a minor issue, but to put it in context, a file that parses in 300 milliseconds on GWT 7, now takes 3000 milliseconds under GWT 2.8.0, and it is directly because of this change.

**Comment by dankurka** [comment]
I am currently out of office but will get to this when I return on the 8th. Adding this to GWT 2.8 milestone for now

**Comment by gkdn** [comment]
The batch approach in the patch is actually the suggested hybrid approach.
Original implementation was using String.fromCharCode.apply
Patch originally suggested to use String.fromCharCode for each character
And I suggested to continue using String.fromCharCode.apply in batches.

At the time [jsperf](http://jsperf.com/string-fromcharcode-test/13) showed the batch approach has the best performance characteristics across browsers but unfortunately we can't see it as jsperf is down. Looking at the patch, performance wise I would expect it to behave similarly with 2.7.

If you can send a patch that shows better performance in all browsers for difference sizes of char array, we can accept that patch.

BTW, StringBuilder doesn't use this function.

**Comment by gkdn** [comment]
Actually if I am reading it correctly, the patch you are referring to actually was already released part of GWT 2.7 so it cannot have anything to do with a performance regression.

**Comment by ainsley** [comment]
I created a test harness, and put up a demo of the issue on my own webspace (sorry, it doesn't look pretty).

There remain performance regressions across ie, and firefox ... but I discovered something very interesting as well regarding Chrome's performance regressions.

Chromes MASSIVE performance regression only seems to occur when loading the GWT application from the local file system. Both Firefox and IE face their performance regressions consistently no matter if served via http:// or file://

As such, I have placed a link to the compiled GWT application as well as web hosted versions of the benchmarks.

**Source:**

https://github.com/ainsley/GwtIssue9383Test1

**2.8.0 RC1 compiled version - prettified**

https://iro-iro.xyz/gwtissue9383test/280rc1/

**2.7.0 compiled version - prettified**

https://iro-iro.xyz/gwtissue9383test/270

**Zipfile containing EXACT same compiled version as those hosted online**

https://iro-iro.xyz/gwtissue9383test/issue9383_local_version.zip

**Summary**

- IE has 2x slowdown (IE 11.0.9600.18205).
- Firefox has 2.5x slowdown (however running a very old version of Firefox so I can use the old GWT plugin).
- Chrome (via file://) has greater then 40x slowdown, Maybe a Chrome bug? Please use the linked zipfile to try the benchmarks locally.
- Chrome (via https://) has no discernible slowdown (or speedup) from going from 2.7.0 to 2.8.0 (Chrome 52).
- Did not try Edge browser as not running Windows 10.

**Comment by gkdn** [comment]
Thanks for the detailed info; it was helpful.

I noticed several issues in code generated by GWT 2.8; some are compiler related issues and some are related to JRE.

1. We started putting a clinit for String and add calls to it everywhere. This seems like a compiler issue. Filed #9387. This doesn't seem like affecting performance much but might increase code size.
2. We introduced an unnecessary cast. Filed #9388. Affects the performance slightly.
3. JsInterop introduces $wnd to qualify types so the generated code looks like $wnd.String.fromCharCode.apply. This seems like the root cause of the Firefox slow down (and the Chrome slowdown when app served via  file:// url). Filed #9389

We will fix all these issues probably by tomorrow.

@dankurka: Can you run perf tests before the release so we could see all these regressions? You can run them via sampling, like one patch per month.

**Comment by rluble** [comment]
The main performance regression has been addressed at #9389.
